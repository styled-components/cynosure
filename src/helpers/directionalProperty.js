// @flow
import { error, warning } from '../internalHelpers/_messageHandler'

const positionMap = ['top', 'right', 'bottom', 'left']

function generateProperty(property: string, position: string) {
  if (!property) return position
  const splitPropertyName = property.split('-')
  splitPropertyName.splice(1, 0, position)
  return splitPropertyName.join('-')
}

function generateStyles(property: string, valuesWithDefaults: Array<?string>) {
  const styles = {}
  for (let i = 0; i < valuesWithDefaults.length; i += 1) {
    if (valuesWithDefaults[i]) {
      styles[generateProperty(property, positionMap[i])] = valuesWithDefaults[i]
    }
  }
  return styles
}

/**
 * The directional property helper enables shorthand for direction based properties. It accepts a property and up to four values that map to top, right, bottom, and left, respectively. You can optionally pass an empty string to get only the directional values as properties. You can optionally pass a null argument for a directional value to ignore it.
 * @example
 * // Styles as object usage
 * const styles = {
 *   ...directionalProperty('padding', '12px', '24px', '36px', '48px')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${directionalProperty('padding', '12px', '24px', '36px', '48px')}
 * `
 *
 * // CSS as JS Output
 *
 * div {
 *   'padding-top': '12px',
 *   'padding-right': '24px',
 *   'padding-bottom': '36px',
 *   'padding-left': '48px'
 * }
 */

function directionalProperty(property: string, ...values: Array<?string>) {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (!property) error('expects a property(string) as its firt parameter. However, you did not provide one.')
    if (!values) error('expects at least one value(string) as its tail. However, you did not provide one.')
    if (values.length > 5) warning(`expects no more than 4 values(string) as its tail. However, you provided ${values.length} instead, ${values.length - 4} of which were ignored.`)
  }
  // $FlowIgnoreNextLine doesn't understand destructuring with chained defaults.
  const [firstValue, secondValue = firstValue, thirdValue = firstValue, fourthValue = secondValue] = values
  const valuesWithDefaults = [firstValue, secondValue, thirdValue, fourthValue]
  return generateStyles(property, valuesWithDefaults)
}

export default directionalProperty
