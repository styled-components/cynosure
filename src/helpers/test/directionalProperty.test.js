// @flow
import directionalProperty from '../directionalProperty'

describe('directionalProperty', () => {
  it('properly generates properties when passed a hyphenated property', () => {
    expect(directionalProperty('border-width', '12px')).toEqual({
      borderBottomWidth: '12px',
      borderLeftWidth: '12px',
      borderRightWidth: '12px',
      borderTopWidth: '12px',
    })
  })

  it('properly generates properties when passed a camelCased property', () => {
    expect(directionalProperty('borderWidth', '12px')).toEqual({
      borderBottomWidth: '12px',
      borderLeftWidth: '12px',
      borderRightWidth: '12px',
      borderTopWidth: '12px',
    })
  })

  it('properly passes just the position when not given a property', () => {
    expect(directionalProperty('', '12px')).toEqual({
      bottom: '12px',
      left: '12px',
      right: '12px',
      top: '12px',
    })
  })

  it('properly sets unitless 0', () => {
    expect(directionalProperty('', 0)).toEqual({
      bottom: 0,
      left: 0,
      right: 0,
      top: 0,
    })
  })

  // One Param
  it('properly applies a value when passed only one', () => {
    expect(directionalProperty('border', '12px')).toEqual({
      borderBottom: '12px',
      borderLeft: '12px',
      borderRight: '12px',
      borderTop: '12px',
    })
  })

  it('properly applies a integer value when passed only one', () => {
    expect(directionalProperty('border', 12)).toEqual({
      borderBottom: 12,
      borderLeft: 12,
      borderRight: 12,
      borderTop: 12,
    })
  })

  // Two Params
  it('properly applies values when passed two', () => {
    expect(directionalProperty('border', '12px', '24px')).toEqual({
      borderBottom: '12px',
      borderLeft: '24px',
      borderRight: '24px',
      borderTop: '12px',
    })
  })
  it('properly applies values when passed two integers', () => {
    expect(directionalProperty('border', 12, 24)).toEqual({
      borderBottom: 12,
      borderLeft: 24,
      borderRight: 24,
      borderTop: 12,
    })
  })
  it('properly applies values when passed a string and an integer', () => {
    expect(directionalProperty('border', 12, '24px')).toEqual({
      borderBottom: 12,
      borderLeft: '24px',
      borderRight: '24px',
      borderTop: 12,
    })
  })
  it('properly skips top and bottom properties when first value is null', () => {
    expect(directionalProperty('border', null, '12px')).toEqual({
      borderLeft: '12px',
      borderRight: '12px',
    })
  })
  it('properly skips left and right properties when second value is null', () => {
    expect(directionalProperty('border', '12px', null)).toEqual({
      borderBottom: '12px',
      borderTop: '12px',
    })
  })

  // Three Params
  it('properly applies values when passed three', () => {
    expect(directionalProperty('border', '12px', '24px', '36px')).toEqual({
      borderBottom: '36px',
      borderLeft: '24px',
      borderRight: '24px',
      borderTop: '12px',
    })
  })
  it('properly skips top property when first value is null', () => {
    expect(directionalProperty('border', null, '24px', '36px')).toEqual({
      borderBottom: '36px',
      borderLeft: '24px',
      borderRight: '24px',
    })
  })
  it('properly skips right and left properties when second value is null', () => {
    expect(directionalProperty('border', '12px', null, '36px')).toEqual({
      borderBottom: '36px',
      borderTop: '12px',
    })
  })
  it('properly skips bottom property when last value is null', () => {
    expect(directionalProperty('border', '12px', '24px', null)).toEqual({
      borderLeft: '24px',
      borderRight: '24px',
      borderTop: '12px',
    })
  })
  it('properly applies values when passed a mixture of three value types', () => {
    expect(directionalProperty('border', 12, '24px', null)).toEqual({
      borderLeft: '24px',
      borderRight: '24px',
      borderTop: 12,
    })
  })

  // Four Params
  it('properly applies values when passed four', () => {
    expect(
      directionalProperty('border', '12px', '24px', '36px', '48px'),
    ).toEqual({
      borderBottom: '36px',
      borderLeft: '48px',
      borderRight: '24px',
      borderTop: '12px',
    })
  })
  it('properly skips top property when first value is null', () => {
    expect(directionalProperty('border', null, '24px', '36px', '48px')).toEqual(
      {
        borderBottom: '36px',
        borderLeft: '48px',
        borderRight: '24px',
      },
    )
  })
  it('properly skips right property when second value is null', () => {
    expect(directionalProperty('border', '12px', null, '36px', '48px')).toEqual(
      {
        borderBottom: '36px',
        borderLeft: '48px',
        borderTop: '12px',
      },
    )
  })
  it('properly skips bottom property when third value is null', () => {
    expect(directionalProperty('border', '12px', '24px', null, '48px')).toEqual(
      {
        borderLeft: '48px',
        borderRight: '24px',
        borderTop: '12px',
      },
    )
  })
  it('properly skips left property when fourth value is null', () => {
    expect(directionalProperty('border', '12px', '24px', '36px', null)).toEqual(
      {
        borderBottom: '36px',
        borderRight: '24px',
        borderTop: '12px',
      },
    )
  })
  it('properly applies valuew when passed a mixture of four value types', () => {
    expect(directionalProperty('border', 12, '24px', 36, null)).toEqual({
      borderBottom: 36,
      borderRight: '24px',
      borderTop: 12,
    })
  })
})
