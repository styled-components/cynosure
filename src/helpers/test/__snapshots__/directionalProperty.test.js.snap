exports[`directionalProperty properly applies a value when passed only one 1`] = `
Object {
  "borderBottom": "12px",
  "borderLeft": "12px",
  "borderRight": "12px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly applies values when passed four 1`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "48px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly applies values when passed three 1`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "24px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly applies values when passed two 1`] = `
Object {
  "borderBottom": "12px",
  "borderLeft": "24px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly generates properties when passed a camelCased property 1`] = `
Object {
  "borderBottomWidth": "12px",
  "borderLeftWidth": "12px",
  "borderRightWidth": "12px",
  "borderTopWidth": "12px",
}
`;

exports[`directionalProperty properly generates properties when passed a hyphenated property 1`] = `
Object {
  "borderBottomWidth": "12px",
  "borderLeftWidth": "12px",
  "borderRightWidth": "12px",
  "borderTopWidth": "12px",
}
`;

exports[`directionalProperty properly passes just the position when not given a property 1`] = `
Object {
  "bottom": "12px",
  "left": "12px",
  "right": "12px",
  "top": "12px",
}
`;

exports[`directionalProperty properly skips bottom property when last value is null 1`] = `
Object {
  "borderLeft": "24px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips bottom property when third value is null 1`] = `
Object {
  "borderLeft": "48px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips left and right properties when second value is null 1`] = `
Object {
  "borderBottom": "12px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips left property when fourth value is null 1`] = `
Object {
  "borderBottom": "36px",
  "borderRight": "24px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips right and left properties when second value is null 1`] = `
Object {
  "borderBottom": "36px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips right property when second value is null 1`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "48px",
  "borderTop": "12px",
}
`;

exports[`directionalProperty properly skips top and bottom properties when first value is null 1`] = `
Object {
  "borderLeft": "12px",
  "borderRight": "12px",
}
`;

exports[`directionalProperty properly skips top property when first value is null 1`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "24px",
  "borderRight": "24px",
}
`;

exports[`directionalProperty properly skips top property when first value is null 2`] = `
Object {
  "borderBottom": "36px",
  "borderLeft": "48px",
  "borderRight": "24px",
}
`;
